In [754]:
import os
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sympy import *
from scipy import *

## 9.1(б) ##

Методом обратной интерполяции найти корень нелинейного
уравнения, используя приведенные таблицы. Оценить точность полученного решения.

$$ (x - 1)^2 - \frac{e^x}{2} = 0 $$

| x    | f(x) |
| --------| ------- |
| 2 | 0.029 |
| 2.5 | -0.080 |
| 2.7 | -0.122 |
| 3 | -0.185 |

### Решение ###

Заметим, что $f(x)$ - убывает, поэтому существует g(f(x))

Суть алгоритма заключается в следующем:

1) Мы ищем многочлен Ньютона и записываем его "вперед"
$$ L(x) = f(x_0) + f(x_0, x_1)(x-x_0) + f(x_0, x_1, x_2)(x-x_0)(x - x_1) + ...$$

2) Подставляем x* = 0, и полученоное значение $L(x*)$ будет являться решением

In [755]:
y = np.array([2, 2.5, 2.7, 3]) * 0.1
x = np.array([0.029, -0.080, -0.122, -0.185])

levels = []
s_x = symbols('x')
expr = y[3]
levels.append(y)
for i in range(1, len(x)):
    level = []
    for j in range(0, len(x) - i):
        coef = (levels[i - 1][j] - levels[i - 1][j+1]) / (x[j] - x[j + i])
        if (j == len(x) - i - 1):
            expr_tmp = coef
            for k in  range(j, len(x) -1):
                expr_tmp *= (s_x - x[k + 1])
            expr += expr_tmp
        level.append(coef)
    levels.append(level)

result = expr.subs({s_x: 0})
print(result)
print(expr)

0.212881200695242
-0.476190476190477*x + (-5.28677630773884e-16*x - 6.44986709544138e-17)*(x + 0.185) + (0.540783556978449*x + 0.0432626845582759)*(x + 0.122)*(x + 0.185) + 0.211904761904762


Получили следующий результата
$$x_{solve} = 0.21288 $$
$$x_{table} = 0.213$$

### Оценим ошибку ###

In [756]:
expr = ((s_x - 1)**2 - exp(s_x) /2)
for i in range(0, len(x)):
    expr = diff(expr, s_x)

fact = len(x)
for k in  range(0, len(x)):
    expr *= (s_x - x[k])

print(expr)
result = expr.subs({s_x: 0})
print(result)

-(x - 0.029)*(x + 0.08)*(x + 0.122)*(x + 0.185)*exp(x)/2
2.61812000000000e-5


Ошибка составила:
$$ \Delta = 2.6 * 10^{-5}$$
$$x_{solve} = 0.21288 $$


## 9.1 ##

Для функции, заданной таблично, вычислить значение интеграла
с использованием указанной формулы. Уточнить полученное значение
интеграла с помощью экстраполяции Ричардсона.

а) формулы Симпсона 

б) формула трапеции

### По формуле Симпсона ###

$$\sum^n_{i = 0} = \frac{(b - a)}{6}(f_i(a) + 4f_i(\frac{(a+b)}{2}) + f_i(b))$$

$$f(x) \sim x^2$$

In [757]:
x = np.array([-1, -0.75, -0.5, -0.25, 0, 0.25, 0.5, 0.75, 1])
y = np.array([-1, -0.14, -0.032, 0.01, 0, 0.002, 0.003, 0.0031, 0.0029])

res:float = 0
i = 0
while i < len(x) - 1:
    res = res + ((x[i + 2] - x[i])/ 6 * (y[i] + 4*[i + 1] + y[i + 2]))
    i = i +2
print(res)

res2 = 0
i = 0
while i < len(x) - 1:
    res2 = res2 + ((x[i + 4] - x[i])/ 6 * (y[i] + 4*[i + 2] + y[i + 4]))
    i = i + 4
print(res2)

[1.24540833 1.24540833 1.24540833 1.24540833]
[1.16715 1.16715 1.16715 1.16715]


Получили результат при h:
$$\int^{1}_{-1} f(x) \approx 1.2454$$

Получили результат при 2h:
$$\int^{1}_{-1} f(x) \approx 1.1672$$

### По формуле Трапеции ###

$$\sum^n_{i = 0} = \frac{(b - a)}{2}(f_i(a) + f_i(b))$$


In [758]:
x = [0, 0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2]
y = [0, 0.028, 0.054, 0.078, 0.1, 0.2, 0.133, 0.145, 0.154]

# x = [0, 0.125, 0.25, 0.375 ,0.5, 0.625, 0.75, 0.875, 1]
# y = [1.0, 0.984615, 0.941176, 0.876712, 0.8, 0.719101, 0.64, 0.566372, 0.5]

res:float = 0
i = 0
while i < len(x) -1:
    res = res + ((x[i + 1] - x[i])/ 2 * (y[i] + y[i + 1]))
    i = i + 1
print(res)

res2 = 0
i = 0
while i < len(x) -1:
    res2 = res2 + ((x[i + 2] - x[i])/ 2 * (y[i] + y[i + 2]))
    i = i + 2
print(res2)

0.20375
0.182


Получили результат при h:
$$\int^{2}_{0} f(x) \approx 0.204$$

Получили результат при 2h:
$$\int^{2}_{0} f(x) \approx 0.182$$

### Уточним полученные результаты при помощи экстрополяции Ричардсона ###

$$I = I^{p}(\frac{h}{2}) + \frac{I^{p}(\frac{h}{2})- I^{p}(h)} {2^p - 1} $$

In [759]:
print(0.20375 + ((0.20375 - 0.182)/(3))) # Дельта для трапеции

0.211


In [760]:
print(1.2454 + ((1.2454 - 1.1672)/(15))) # Дельта для Симпсона

1.2506133333333334


### C поправкой Ричардсона ###

1) Симпсон
$$\int =  I + \Delta = 1.2506$$
2) Трапеция
$$\int =  I + \Delta = 0.211$$

## 10.1 ##

Выписать формулы Эйлера с пересчетом для следующих задач:

а) $y′ = x + cos y; \;\; y(1) = 30; \;\;1 ≤ x ≤ 2$,

б) $y′ = x^2 + y^2; \;\; y(2) = 1; \;\; 1 ≤ x ≤ 2$.

Провести вычисления с шагом h = 0,01.

###  Описание ###

При данном подходе рекуррентное соотношение видоизменяется, а именно, вместо $f(x_i, y_i)$ берут среднее арифметическое между $f(x_i, y_i)$ и $f(x_{i+1}, y_{i+1})$.

$$ y_{i+1} = y_i = \frac{h}{2}(f(x_{i}, y_{i}) + f(x_{i+1}, y_{i+1})) $$

Это неявная схема. Она реализуется в две итерации: сначала находится первое приближение, считая $y_i$ начальной

$$Y_{i+1} = y_i + hf(x_i, y_i)$$

затем подставляется в правую часть  вместо $y_{i+1}$
$$ y_{i+1} = y_i + \frac{h}{2}(f(x_{i}, y_{i}) + f(x_{i+1}, Y_{i+1})) $$


In [761]:
def solve(x_i, y_i, a, b, h, expr):
    Y = val = 0
    if x_i > a:
        step = -h
    else:
        step = h

    while x_i <= b and x_i >= a:
        print(x_i)
        Y = float(y_i + h * expr.subs({s_x: x_i, s_y: y_i}))
        val = float((expr.subs({s_x: x_i, s_y: y_i}) + expr.subs({s_x: x_i, s_y: Y})))
        # print(x_i, y_i)
        y_i = float(y_i + h/2 * val)
        x_i = x_i + step

    return y_i


s_x, s_y = symbols('x y')

expr = s_y - 2*s_x / s_y
print(solve(0, 1, 0, 1, 0.2, expr))

expr = s_x + cos(s_y)
print(solve(1, 30, 1, 2, 0.01, expr))

expr = s_x**2 + s_y**2
print(solve(2, 1, 1, 2, 0.01, expr))

# print(y_i)


0
0.2
0.4
0.6000000000000001
0.8
1.0
2.265787377244199
